// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.afsgrpc";
option java_outer_classname = "AfsGrpcProto";
option objc_class_prefix = "AfsG";

package afsgrpc;

service FileService {
  rpc DownloadFile (FileRequest) returns (FileReply) {}
  rpc GetAttribute (FileRequest) returns (AttributeReply) {}
  rpc MakeDir (MakeDirRequest) returns (CreateReply) {}
  rpc UploadFile(UploadRequest) returns (CreateReply) {}
  rpc ReadDir(FileRequest) returns (ReadDirReply) {}
}

message FileRequest {
  string file_path = 1;
}

message FileReply {
  string file_data = 1;
  int32 err = 2;
}

message AttributeReply {
  int64 dev = 1;
  int64 ino = 2;
  int32 mode = 3;
  int64 nlink = 4;
  int32 uid = 5;
  int32 gid = 6;
  int64 rdev = 7;
  int64 size = 8;
  int64 atime = 9;
  int64 mtime = 10;
  int64 ctime = 11;
  int64 blksize = 12;
  int64 blocks = 13;
  int32 res = 14;
}

message MakeDirRequest {
  string dir_path = 1;
  int32 mode = 2;
}

message UploadRequest {
  string file_data = 1;
  string file_path = 2;
}

message CreateFileRequest {
  string file_path = 1;
  int32 mode = 2;
}

message CreateReply {
 int32 err = 1;
}

message ReadDirReply {
  repeated int64 inodenumber = 1;
  repeated int32 type = 2;
  repeated string name = 3;
  int32 err = 4;
}